src/core/apparmor-setup.c

```diff
@@ -0,0 +1,98 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "macro.h"
+#include "apparmor-setup.h"
+#include "apparmor-util.h"
+#include "string-util.h"
+
+#if HAVE_APPARMOR
+#include <sys/apparmor.h>
+#endif
+
+#define APPARMOR_PROFILE_NAME "systemd"
+#define APPARMOR_PROFILE_ETC_PATH "/etc/systemd/profile.apparmor"
+#define APPARMOR_PROFILE_USR_PATH "/usr/lib/systemd/profile.apparmor"
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r = -1;
+        _cleanup_close_ int fd = -1;
+        aa_kernel_interface *interface = NULL;
+        const char* profile_file = APPARMOR_PROFILE_ETC_PATH;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        r = read_one_line_file("/proc/self/attr/apparmor/current", &current_profile);
+        if (r < 0) {
+                if (errno == ENOENT) {
+                        r = read_one_line_file("/proc/self/attr/current", &current_profile);
+                        if (r < 0) {
+                                log_warning_errno(errno, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/current: %m");
+                                return 0;
+                        }
+                } else {
+                        log_warning_errno(errno, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/apparmor/current: %m");
+                        return 0;
+                }
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("Systemd is already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        if (access(profile_file, F_OK) < 0) {
+                profile_file = APPARMOR_PROFILE_USR_PATH;
+                if (access(profile_file, F_OK) < 0) {
+                        log_debug("AppArmor profile file neither found at path " APPARMOR_PROFILE_ETC_PATH " nor " APPARMOR_PROFILE_USR_PATH ".");
+                        return 0;
+                }
+        }
+
+        fd = open(profile_file, O_RDONLY|O_CLOEXEC);
+        if (fd < 0)  {
+                log_warning_errno(errno, "Failed to open binary AppArmor profile file %s: %m", profile_file);
+                return 0;
+        }
+
+        r = aa_kernel_interface_new(&interface, NULL, NULL);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to get a new AppArmor kernel interface: %m");
+                return 0;
+        }
+
+        r = aa_kernel_interface_load_policy_from_fd(interface, fd);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to load AppArmor profile from file %s into the kernel: %m", profile_file);
+                aa_kernel_interface_unref(interface);
```

Comment:
instead of unreffing this explicitly in each exit path, please use DEFINE_TRIVIAL_CLEANUP_FUNC() and then use `_cleanup_(…)` on the variable to automatically destroy the object when the function is left.